## Dryad is a pipeline to construct a reference free core-genome or SNP phylogenetic tree for examining prokaryote relatedness in outbreaks.
[![Build Status](https://travis-ci.org/k-florek/dryad.svg?branch=master)](https://travis-ci.org/k-florek/dryad)

Dryad is a pipeline that uses several programs and pipelines to construct trees from either a core set of genes or a set of SNPs from a reference genome. The pipeline uses (nextflow)[http://nextflow.io] to efficiently use compute resources regardless of the environment increasing the speed and portability of the pipeline. The pipeline uses Docker containers to maintain the stability, reproducibility, and portability by keeping the applications and pipelines used by this pipeline in controlled environments. This also reduces issues surrounding the installation of dependencies.

### Table of Contents:
[Usage](#Using-the-pipeline)  
[Core-genome](#Core-Genome-phylogenetic-tree-construction)  
[SNP](#SNP-phylogenetic-tree-construction)  
[Output](#Output-files)  
[Dependencies](#Dependencies)  

#### Using the pipeline

Dryad is designed to work with paired end raw Illumina short reads. The pipeline only requires a directory that contains all of the raw reads.

Note: Dryad by default splits names using an `_`. For example, `sample-1_R1_L001.fastq.gz` will be named `sample-1` throughout the pipeline. If you use `_` in your sample names please update the `params.name_split_on="_"` in the `dryad.config` file to what you would like to use.

```
usage: dryad [-h] [--core-genome] [--snp] [-ar] [-r <path>] [-t cpus]
             [-m memory in GB]
             reads_path

A comprehensive tree building program.

positional arguments:
  reads_path          Path to the location of the raw reads in the fastq
                      format.

optional arguments:
  -h, --help          show this help message and exit
  --core-genome, -cg  Construct a core-genome tree.
  --snp, -s           Construct a SNP tree. Note: Requires a reference genome
                      in fasta format (-r).
  -ar                 Detect AR mechanisms.
  -r <path>           Reference genome for SNP pipeline.
  -t cpus             Maxmium number of cpus to use, default 8.
  -m memory in GB     Maxmium number of GB of memory to use, default 16.
```

Dryad begins with a quality trimming step to trim the reads of low quality bases at the end of the read using (Trimmomatic)[http://www.usadellab.org/cms/?page=trimmomatic] v0.39. After the trimming process Phix and Adapter Sequences are removed using (BBTools)[https://jgi.doe.gov/data-and-tools/bbtools/] v38.76 the read information is then used by each pipeline as needed.

#### Core Genome phylogenetic tree construction
The core-genome pipeline takes the trimmed reads and generates a phylogenetic tree that can be used for inferring outbreak relatedness. The pipeline only requires the text file containing the read paths mentioned above. This pipeline is based loosely off of the pipeline described here by [Oakeson et. al](https://www.ncbi.nlm.nih.gov/pubmed/30158193).

The pipeline uses the following applications and pipelines:

Shovill v1.0.4 (https://github.com/tseemann/shovill)
Shovill is a pipeline centered around SPAdes but alters some of the steps to get similar results in less time.

Prokka v1.14.0 (https://github.com/tseemann/prokka)
Prokka is a whole genome annotation tool that is used to annotate the coding regions of the assembly.

Roary v3.6.0 (https://github.com/sanger-pathogens/Roary)
Roary takes the annotated genomes and constructs a core gene alignment.

IQ-Tree v1.6.7 (http://www.iqtree.org/)
IQ-Tree uses the core gene alignment and creates a maximum likelihood phylogenetic tree bootstraped 1000 times.

#### SNP phylogenetic tree construction
The SNP pipeline takes the trimmed reads and generates a phylogenetic tree that can be used for inferring outbreak relatedness. The pipeline requires the text file containing the read paths mentioned above and a reference genome in a fasta file format.

CFSAN SNP Pipeline v2.0.2 (https://github.com/CFSAN-Biostatistics/snp-pipeline)

IQ-Tree v1.6.7 (http://www.iqtree.org/)
IQ-Tree uses an alignment of the SNP sites to create a maximum likelihood phylogenetic tree bootstraped 1000 times.


#### Output files
The dryad pipeline creates a folder named `dryad-XXXXXXXXXX` when it is initially run. The x values are replaced with datetime information. In the dryad folder you will find the intermediate and result files. The following files are the key output files generated by the pipeline.

**core_genome_tree.tree** - the core-genome phylogenetic tree created by the core-genome pipeline  
**snp_tree.tree** - the SNP tree created by the SNP pipeline  
**snp_distance_matrix.tsv** - the SNP distances generated by the SNP pipeline  

#### Dependencies
Python 3 (https://www.python.org)  
Docker (https://www.docker.com) or Singularity (https://sylabs.io/singularity/)  
Java 8 (https://openjdk.java.net/)  
